src/Data/List.purs

```diff
@@ -399,27 +398,17 @@ filter p = go Nil
 -- | powerSet :: forall a. [a] -> [[a]]
 -- | powerSet = filterM (const [true, false])
 -- | ```
-filterM :: forall a m. (Monad m) => (a -> m Boolean) -> List a -> m (List a)
-filterM _ Nil = return Nil
-filterM p (Cons x xs) = do
-  b <- p x
-  xs' <- filterM p xs
-  return $ if b
-           then Cons x xs'
-           else xs'
+filterM :: forall a m. (Applicative m) => (a -> m Boolean) -> List a -> m (List a)
+filterM _ Nil = pure Nil
+filterM p (Cons x xs) = consIf <$> p x <*> filterM p xs
+  where consIf b xs' = if b then Cons x xs' else xs'
 
 -- | Apply a function to each element in a list, keeping only the results which
 -- | contain a value.
 -- |
 -- | Running time: `O(n)`
 mapMaybe :: forall a b. (a -> Maybe b) -> List a -> List b
-mapMaybe f = go Nil
-  where
-  go acc Nil = reverse acc
-  go acc (Cons x xs) =
-    case f x of
-      Nothing -> go acc xs
-      Just y -> go (Cons y acc) xs
+mapMaybe f = foldMap (maybe Nil singleton <<< f)
```

Comment:
I believe it should be the same as the above functions using `foldMap`, since `foldMap` is right fold, and concatenation to the singleton should be constant time. It does feel like a fair amount of wrapping/unwrapping, perhaps it would be better to define `mapMaybe` (and use it for filter as well)?